models/wrf/model_mod.f90

```diff
@@ -694,6 +694,19 @@ subroutine static_init_model()
 
    ! JPH now that we have the domain ID just go ahead and get type indices once
    ! NOTE: this is not strictly necessary - can use only stagger info in the future (???)
+
+   ! If using WRF temperature variable 'T' instead of 'THM' fail immediately to 
+   ! prevent boundscheck errors. For WRFv4 and later version variable 'T' is 
+   ! diagnostic, thus updating the 'THM' variable (prognostic) is preferred.
+
+   if ( wrf%dom(id)%type_t >= 0 .and. get_type_ind_from_type_string(id,'T') >=0) then
```

Comment:
so at the comment, this check isn't checking that THM is in state and type_t

Reply:
> so at the comment, this check isn't checking that THM is in state and type_t

I went back and forth on this -- yes, the code will error out if user includes 'T' as temperature variable.  It does not error out if 'THM' is in the DART state, but does give that information in the error message.

I could expand the error by saying:

`if ( wrf%dom(id)%type_t >= 0 .and. (get_type_ind_from_type_string(id,'T') >=0 .or. get_type_ind_from_type_string(id,'THM') <=0)) then`

Is that preferable?

Reply:
Suggested edit will now error out if THM is *not* in DART state but type_t WRF variable is declared.

Reply:
line 702:
`if ( wrf%dom(id)%type_t >= 0 .and. get_type_ind_from_type_string(id,'T') >=0)` then

`wrf%dom(id)%type_t` isn't defined at this stage (unless I'm missing something) so this check `wrf%dom(id)%type_t >= 0` is using an uninitialized value for type_t

line 713 is where wrf%dom(id)%type_t is set.
`wrf%dom(id)%type_t      = get_type_ind_from_type_string(id,'THM')`

Reply:
so to back up a bit, what has to be true to run wrf-dart?

THM has to be in the state and type_t has to be THM?

What's not allowed?
T can not be in the state vector?
T can be in the state vector but can not be type_t?

Reply:
> so to back up a bit, what has to be true to run wrf-dart?
> 
> THM has to be in the state and type_t has to be THM?
> 
> What's not allowed? T can not be in the state vector? T can be in the state vector but can not be type_t?

The only thing that is not allowed is having the WRF variable 'T' in the DART state because it can throw the 'boundscheck' error which gives a non-intuitive error as described in #728 .  There are other secondary reasons to not include 'T' in WRF -- because its diagnostic in WRFv4+, not prognostic, so updating it will not improve the forecast.  Users who are used to older WRF versions may not be aware of this.

I also believe that including a 'type_t' WRF variable in the DART state is mandatory because of the way the WRF model_mod.f90 works.   Even if you don't intend the 'type_t' variable to be updated by DART, the code still does boundchecks on the type_t variable for many (all?) WRF-DART simulations.

Given the 'type_t' requirement, and secondly, that using variable 'T' is not permitted, this makes including 'THM' mandatory because it is the only replacement variable that I know of for 'T' (3D temperature).